From 0073887e8e0d83de47f25cab9945ce0d8d63f43c Mon Sep 17 00:00:00 2001
From: Roxana Nicolescu <rnicolescu@ciq.com>
Date: Mon, 1 Dec 2025 12:22:59 +0100
Subject: [PATCH] wifi: mac80211: check S1G action frame size

jira VULN-154631
cve CVE-2023-53257
commit-author Johannes Berg <johannes.berg@intel.com>
commit 19e4a47ee74718a22e963e8a647c8c3bfe8bb05c

Before checking the action code, check that it even
exists in the frame.

	Reported-by: syzbot+be9c824e6f269d608288@syzkaller.appspotmail.com
	Signed-off-by: Johannes Berg <johannes.berg@intel.com>
(cherry picked from commit 19e4a47ee74718a22e963e8a647c8c3bfe8bb05c)
	Signed-off-by: Roxana Nicolescu <rnicolescu@ciq.com>
Signed-off-by: Roxana Nicolescu <rnicolescu@ciq.com>
---
 net/mac80211/rx.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 71972f2bd2ae..1d94e3ff0dfc 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3551,6 +3551,10 @@ ieee80211_rx_h_action(struct ieee80211_rx_data *rx)
 			break;
 		goto queue;
 	case WLAN_CATEGORY_S1G:
+		if (len < offsetofend(typeof(*mgmt),
+				      u.action.u.s1g.action_code))
+			break;
+
 		switch (mgmt->u.action.u.s1g.action_code) {
 		case WLAN_S1G_TWT_SETUP:
 		case WLAN_S1G_TWT_TEARDOWN:
-- 
2.47.3

